### PR TITLE
RFC: Client Controlled Nullability Operators Implementation

### PR DESCRIPTION
**Pull request by twof**
Implementation of https://github.com/graphql/graphql-spec/issues/867 along with the suggested `?` syntax. 

[Here's a doc with a list of some of the important test cases and the justifications for their behavior.](https://docs.google.com/document/d/1relgXUdhJrlaSR70lrITQKmcIHXjIVyWxDCKRuzMEvs/edit#heading=h.gm7m1frqgtc4)

This PR includes work done by @xuewei8910, @aprilrd, @magicmark, and @lizjakubowski.

### CODE REVIEW

File: package.json
Comment (twof) [inline package.json:66]: I think this can be removed

File: package-lock.json
Comment (twof) [inline package-lock.json:28]: We can revert changes to this file

**Comment by twof** [comment]
> [CLA Missing ID](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/12261526/38307428/3281/#/?version=2)
> 
> * ❌ The commit ([43e9fd1](https://github.com/graphql/graphql-js/commit/43e9fd1334db7ea0a10d020938b4cc8b61cbda7a) ,c0e1e7b63d0b95ad00e5b5089d6f7cc1aafe6ca7 ,d63f290dde75c0c80352461947d8f09855430e22 ,8d78724e8569db4f173229981a61325e28d2d5e3 ,08f9319f07fadcc5abbace5b839ccf86f735ef41 ,043f1071dbe04d63fdb31896a7b9e683ade575ba ,102586d45f7f1cee51a9bcbf4595eaf5c18ad4c0 ,301a4ab4fa01395cf17b4c74e84f73124bd9f7ab ,2d7f496ae681e2dee7381eb04c76dc5ca4a44800) is missing the User's ID, preventing the EasyCLA check. [Consult GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve.For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

All of these commits are mine, but they weren't attributed to any github account for one reason or another.

**Comment by saihaj** [comment]
@twof try rebasing and overwriting author on commits https://stackoverflow.com/a/54363151/11321732

**Comment by twof** [comment]
@saihaj Thanks! That worked.

**Comment by IvanGoncharov** [comment]
@twof I enabled CI, and some tests are failing.

File: src/validation/rules/RequiredStatusOnFieldMatchesDefinitionRule.ts
Comment (twof) [inline src/validation/rules/RequiredStatusOnFieldMatchesDefinitionRule.ts:33]: cleanup. Comment is cargo culted

File: src/utilities/__tests__/TypeInfo-test.ts
Comment (twof) [inline src/utilities/__tests__/TypeInfo-test.ts:15]: I might need a test in this file somewhere

File: src/validation/rules/OverlappingFieldsCanBeMergedRule.ts
Comment (twof) [inline src/validation/rules/OverlappingFieldsCanBeMergedRule.ts:615]: Is it okay not handling errors here if I expect the errors these functions throw to be handled in a different rule? Is there any reason to think one rule will be run and the other won't?

File: src/validation/rules/RequiredStatusOnFieldMatchesDefinitionRule.ts
Comment (twof) [inline src/validation/rules/RequiredStatusOnFieldMatchesDefinitionRule.ts:17]: Update this link once I've added this rule to the spec

File: src/language/printer.ts
Comment (twof) [inline src/language/printer.ts:95]: Typescript thinks that `required` is a `string`, but it's actually a `ComplexRequiredStatus`. I'm not sure why that's happening, so this is how I'm dealing with it. Let me know if there's something else I can do.

**Comment by twof** [comment]
Hello again! I've implemented the list syntax described here: https://github.com/graphql/graphql-spec/pull/895#issuecomment-963909931

I'd like a code review on this before I add the list syntax to the spec, in case the implementation ends up being dramatically different after I apply any feedback I get. It's a large PR, so if there's any way I can make reviewing this easier, I'd be happy to do so.

File: src/index.ts
Comment (IvanGoncharov) [inline src/index.ts:240]: this is strange.
Problems with rebase?

**Review by IvanGoncharov — CHANGES_REQUESTED** [review]

File: src/language/__tests__/lexer-test.ts
Comment (IvanGoncharov) [inline src/language/__tests__/lexer-test.ts:171]: please extract changes in this file as separate PR, happy to merge them as is.
This PR is quite big so let's try to extract parts that don't require spec changes.
It will review simpler for everyone from WG.

Also please add test for `?` similar to what we already have for `!`

**Review by IvanGoncharov — CHANGES_REQUESTED** [review]

File: src/index.ts
Comment (twof) [reply src/index.ts:240]: I think my editor was telling me these were deprecated, but I can try bringing them back.

**Comment by twof** [comment]
Extracted changes that don't need to be in this PR:
https://github.com/graphql/graphql-js/pull/3376

File: src/language/__tests__/lexer-test.ts
Comment (twof) [reply src/language/__tests__/lexer-test.ts:171]: I think I've extracted everything I can. Unfortunately it's not a whole lot, but let me know if I've missed anything. I've also added `?` tests to `lexer-test.ts`.

https://github.com/graphql/graphql-js/pull/3281#issuecomment-975831116

Comment (twof) [reply src/language/__tests__/lexer-test.ts:171]: Let me know if there are any notes I can write up that might help you and others review this PR. I fully understand that it's quite large!

File: src/language/ast.ts
Comment (IvanGoncharov) [inline src/language/ast.ts:388]: This should be a part of AST.
Without it you can't point to error.
For example, I wrote query like that:
```
{ 
   twoDimentionalArray[[[!]!]!]!
}
```
So error should be highlighted as follow:
```
{ 
   twoDimentionalArray[[[!]!]!]!
                        ^^^
}
```
If we go with "support array" root, then we should introduce 3 new AST nodes:
1. for `!`
2. for `?`
3. for `[]`

Also it will allow making property in `FieldNode` optional to represent "normal" state.
```
readonly requireModifiers?: RequireModifierNode;
```

**Review by IvanGoncharov — CHANGES_REQUESTED** [review]

File: src/language/ast.ts
Comment (IvanGoncharov) [reply src/language/ast.ts:388]: Also it will make `print` implementation simpler and make this feature compatible with other AST based function like `visit`.

**Comment by IvanGoncharov** [comment]
> Let me know if there are any notes I can write up that might help you and others review this PR. I fully understand that it's quite large!

@twof Great idea. 
Let's do first round of review where I check "architectural stuff", e.g. AST design.
This round can provoke quite big changes so I suggest writing notes after it finishes.

**Comment by twof** [comment]
@IvanGoncharov Sounds like a plan. Thanks!

File: src/language/ast.ts
Comment (IvanGoncharov) [reply src/language/ast.ts:388]: Thinking more about it, I'm not sure if how to design AST or if we should have 3 new nodes or one more complex node. 
But 100% sure it should be implemented as recursive AST nodes.
So please feel free to experiment with design and see what will fit our AST design.

**Comment by twof** [comment]
@IvanGoncharov  I'll take a look at those failing jobs in the morning, but `npm test` is passing on my machine. In the meantime if you'd like to look at the changes I've made and let me know what you think, the architecture is ready for review again.

I'll go over everything and do some cleanup in the morning, so don't worry too much about the details for now. Also let me know if there are any portions you'd like me to write up explanations for.

Summary of changes: 
- `ComplexRequiredStatus` has been broken up into three `Node` interfaces.
- Printing and parsing is now much simpler. You were totally right about that.
- `modifiedOutputType` now also uses a visitor.

Todo:
- It sounds like from the spec feedback that people want the list nullability modifier syntax (`[!]!`) to always match the depth of the type definition for the field it's being applied to. It currently fails if the modifier is too deep (`[[]]`), but it still succeeds if it's too shallow (`!`).
- People haven't landed on behavior for the question mark syntax yet, but that's liable to cause significant changes to this implementation.

**Comment by twof** [comment]
@IvanGoncharov Hey Ivan. During the meeting last Thursday you mentioned that you'd be open to releasing this feature behind an experimental flag while it's still in review, and you had some other features that did something similar in the past. Are you able to point me towards an example of a feature that used an experimental flag? I can implement something similar as part of this PR if you're still interested in doing that.

**Comment by saihaj** [comment]
Hey @twof we currently publish `defer and stream` as experimental flag you can read more here: https://github.com/graphql/graphql-js#experimental-features. I think for this to happen we first need to get your this branch to the repo. So @IvanGoncharov can help you give rights and set up a branch here which you can push to. Then each time there are changes we just publish a new experimental tag example: https://github.com/graphql/graphql-js/releases/tag/v16.0.0-experimental-stream-defer.5. 

But I think this approach really isn't working well and @IvanGoncharov wanted to merge everything to `main` so we can have unstable `main` and have canary releases. But as per JS-WG it was suppose to happen last Monday or Wednesday (as per a comment I saw) but no updates on that end. But hopefully we can get something this week and have this merged soon cause I want to play around with this feature :) cause I don't want to build myself or start publishing on my fork.

**Comment by saihaj** [comment]
> but npm test is passing on my machine

I assume you are on node `v16` cause `v16` tests are the only ones passing.

File: src/utilities/applyRequiredStatus.ts
Comment (saihaj) [inline src/utilities/applyRequiredStatus.ts:72]: Might wanna change this cause it is only available in node `v16` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at and we want to support `v12`+

**Review by saihaj — CHANGES_REQUESTED** [review]

File: src/utilities/applyRequiredStatus.ts
Comment (twof) [reply src/utilities/applyRequiredStatus.ts]: Thanks for pointing that out!

**Comment by twof** [comment]
@saihaj Great! Let me know if there's anything I can do to help with the experimental flag setup, or if yall have anything else I can fix up about the PR. It looks like most checks are passing.

**Comment by IvanGoncharov** [comment]
@twof I suggest moving in steps.
As the first step let's merge support for `!` and `?` in AST (`parse`, `print`, AST types, `visit`, etc.).
Please split out a smaller PR with just AST changes similar to this one https://github.com/graphql/graphql-js/pull/1141 (note it's more than 4 years old, so code changed a bit since then).
I think one experimental flag for both `!` and `?` would be enough.
Also please make a short summary of the discussion around `!` and `?` syntax (only syntax) and open questions if any and add them in the commit message (or PR message).

That way we will merge noncontroversial stuff first and focus on stuff that needs disscussion.
Also, it will unblock your work with `apollo-ios` and `graphql-codegen`.

**Comment by twof** [comment]
@IvanGoncharov Here's the PR with the stuff you asked for: https://github.com/graphql/graphql-js/pull/3418

File: src/language/ast.ts
Comment (michaelstaib) [inline src/language/ast.ts:397]: Why is this called SupportArrayNode instead of ListNullabilityNode or at least SupportListNode? In GraphQL we are always talking about lists, not arrays.

File: src/language/ast.ts
Comment (michaelstaib) [inline src/language/ast.ts:386]: Why are the node name and the kind name diverging?

File: src/language/ast.ts
Comment (twof) [reply src/language/ast.ts:397]: The name was based on this comment: https://github.com/graphql/graphql-js/pull/3281#discussion_r754859308

I'm not really a fan of the name I picked either, and I'm down to change it before everything gets merged. `ListNullabilityNode` sounds good to me, so I'll go with that.

File: src/language/parser.ts
Comment (michaelstaib) [inline src/language/parser.ts:446]: Shouldn't this come before the selection set and before the directives?

`Field : Alias? Name Arguments? Nullability? Directives? SelectionSet?`


The kitchen sink test never tests a field with selection set:

```
      field3!
       requiredField4: field4!
       field5?
       optionalField6: field6?
       unsetListItemsRequiredList: listField[]!
       requiredListItemsUnsetList: listField[!]
       requiredListItemsRequiredList: listField[!]!
       unsetListItemsOptionalList: listField[]?
       optionalListItemsUnsetList: listField[?]
       optionalListItemsOptionalList: listField[?]?
       multidimensionalList: listField[[[!]!]!]!
```

This must also be tested in its various combinations
```
{
  field? {
    name
  }
}
```

Comment (twof) [reply src/language/parser.ts:446]: Yeah good point. I'll put up some more tests

**Comment by twof** [comment]
Hey @michaelstaib! Thanks for the review. I'm in the process of breaking up this PR into smaller changesets. I've addressed your feedback for the `?` and `!` portions of the proposal in https://github.com/graphql/graphql-js/pull/3418 and your feedback related to the list syntax in https://github.com/twof/graphql-js/pull/2.

@IvanGoncharov @saihaj In order to reduce confusion, would it be helpful to close this PR and designate https://github.com/graphql/graphql-js/pull/3418 as the canonical implementation branch for this RFC?

File: src/language/parser.ts
Comment (michaelstaib) [reply src/language/parser.ts:446]: I have fixed this issue in our implementation and have tests now for this.

It must be like the following:

```
return this.node<FieldNode>(start, {
      kind: Kind.FIELD,
      alias,
      name,
      arguments: this.parseArguments(false),
      required: this.parseRequiredStatus(),
      directives: this.parseDirectives(false),
       selectionSet: this.peek(TokenKind.BRACE_L)
         ? this.parseSelectionSet()
         : undefined
```
       
At the moment it would break in your implementation when it is combined with a selection set or if you combine tests with directives.


Here is our implementation of this ... its c# but the order is correct.
https://github.com/ChilliCream/hotchocolate/blob/b1243030b9e20f6c4a22780857e5cd2031d16c68/src/HotChocolate/Language/src/Language.Utf8/Utf8GraphQLParser.Operations.cs#L255

Comment (twof) [reply src/language/parser.ts:446]: Yep! That's what I went with too

https://github.com/graphql/graphql-js/pull/3418/files#diff-297a41e54c25bbd5c846bb55d7acaaca9f2c9f5fa654b19004da4efbf9a88154R463

File: src/execution/execute.ts
Comment (twof) [inline src/execution/execute.ts:495]: I don't like this pattern. Reassess.

**Pull request closed without merging**
